### PR TITLE
Refactor feature flag overrider architecture to make use of the manager pattern

### PR DESCRIPTION
**Pull request by jeremynikolic**

### CODE REVIEW

File: src/Overriders/InMemoryOverrider.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+namespace Worksome\FeatureFlags\Overriders;
+
+use Illuminate\Support\Arr;
+use Worksome\FeatureFlags\Contracts\FeatureFlagEnum;
+use Worksome\FeatureFlags\Contracts\FeatureFlagOverrider;
+
+class InMemoryOverrider implements FeatureFlagOverrider
+{
+
+    /**
```
Comment (owenvoke) [inline src/Overriders/InMemoryOverrider.php:14]:
```suggestion
    /**
```

**Review by owenvoke — COMMENTED** [review]
Looks great in my opinion. 👍🏻 Would be good to do some testing platform with a PR to ensure everything transitions smoothly.

Only things I've found should be resolved with an `ecs --fix` (except the comment I made, which needs to be done manually).

**Comment by jeremynikolic** [comment]
@owenvoke Platform would require small adjustments but I was planning to do a PR there.
What would be best for that ? 
should I create a "test" tag pointing on this PR and have the platform PR point to it for testing ?

**Comment by owenvoke** [comment]
@jeremynikolic, if you add [the branch as the version](https://packagist.org/packages/worksome/feature-flags#dev-feat_overrider_refactor), it should all work fine. 👍🏻
```json
...
"worksome/feature-flags": "dev-feat_overrider_refactor"
...
```

**Comment by jeremynikolic** [comment]
@owenvoke wonderful, I did not know you could do that :D

File: src/FeatureFlagsOverriderManager.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        return new ConfigOverrider(
+            $this->config,
+        );
+    }
+
+    public function createInMemoryDriver(): InMemoryOverrider
+    {
+        /** @var bool|null $overrideAll */
+        $overrideAll = $this->config->get('feature-flags.overriders.in-memory.override-all');
```
Comment (owenvoke) [inline src/FeatureFlagsOverriderManager.php:24]: I think it's safe to remove this doc comment, as we already check below for if it's not `bool` or `null`. 👍🏻
```suggestion
        $overrideAll = $this->config->get('feature-flags.overriders.in-memory.override-all');
```

File: src/Overriders/InMemoryOverrider.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+     * @param array<string, bool|null> $overrides
+     * @param bool|null $overrideAll
+     */
+    public function __construct(private array $overrides = [], private bool|null $overrideAll = null)
```
Comment (jeremynikolic) [inline src/Overriders/InMemoryOverrider.php:18]: @owenvoke I would love your opinion here 🕵️ 

I've set it so this InMemory provider may be instantiated with "default" values coming from the config.
But I am unsure if this makes a lot of sense or not 🤔 
(I confess I was having fun with the manager pattern when implementing it 😄 )

File: src/FeatureFlagsServiceProvider.php [hunk 4]
```diff
@@ -50,10 +50,9 @@ public function register(): void
             ]);
         });
 
-        $this->app->singleton(
-            FeatureFlagsManager::class,
-            static fn (Container $container) => new FeatureFlagsManager($container)
-        );
+        $this->app->singleton(FeatureFlagsManager::class,);
+
+        $this->app->singleton(FeatureFlagsOverriderManager::class,);
```
Comment (owenvoke) [inline src/FeatureFlagsServiceProvider.php:55]:
```suggestion
        $this->app->singleton(FeatureFlagsManager::class);

        $this->app->singleton(FeatureFlagsOverriderManager::class);
```

File: src/FeatureFlagsApiManager.php [hunk 5]
```diff
@@ -14,9 +14,8 @@ class FeatureFlagsApiManager extends Manager
     public function createLaunchDarklyDriver(): LaunchDarklyApiProvider
     {
         $token = $this->config->get('feature-flags.providers.launchdarkly.access-token');
-        if (! is_string($token)) {
-            throw new LaunchDarklyMissingAccessTokenException();
-        }
+
+        assert(is_string($token), new LaunchDarklyMissingAccessTokenException());
```
Comment (owenvoke) [inline src/FeatureFlagsApiManager.php:18]: Just to say, `assert()` only throws in development environments. We probably want to stick with the previous way? 🤔

File: src/Overriders/InMemoryOverrider.php (hunk 3 above)
Comment (owenvoke) [reply src/Overriders/InMemoryOverrider.php:18]: Hmm, personally I don't think it makes sense to have it in the config. 🤔 I'd say that for usage in tests, etc. (where this is likely to be used), it should be configured with the new `->set()` or `->setAll()` methods. 🤷🏻

However, I'm not entirely sure, as I'm also not a huge fan of having `->set()` / `->setAll()` on the interface. 🤔

File: src/FeatureFlagsApiManager.php (hunk 5 above)
Comment (jeremynikolic) [reply src/FeatureFlagsApiManager.php:18]: Oooh ok I did not know they only throw in dev.
Too bad, I liked the syntax of it 😅

File: src/Overriders/InMemoryOverrider.php (hunk 3 above)
Comment (jeremynikolic) [reply src/Overriders/InMemoryOverrider.php:18]: Yeah I think I agree about the InMemoery, I would remove the default from config part.


I started adding the `set`and `setAll` methods as it made sense being able to swap the override and not have to change the code.
Like the overrider already knows where to look for values so it kind of makes sense for it to know how to store them as well 🤔 
This way the overrider has full responsibility around it getting and setting those 🤔

File: src/FeatureFlagsApiManager.php (hunk 5 above)
Comment (jeremynikolic) [reply src/FeatureFlagsApiManager.php:18]: As we remove the settings of InMemoery default, those just go away 👍

File: src/Overriders/InMemoryOverrider.php (hunk 3 above)
Comment (owenvoke) [reply src/Overriders/InMemoryOverrider.php]: Yeah. I guess it just seemed a bit odd setting config values from outside of the `ConfigRepository` or the config 🤷🏻 But I'm happy whichever way.

**Comment by jeremynikolic** [comment]
@owenvoke I'm waiting for one last run of dusk as I should have addressed everything 🤔

**Comment by jeremynikolic** [comment]
@owenvoke do you feel we are in a good place now with this and the PR to platform to merge ? 🤔

**Comment by owenvoke** [comment]
@jeremynikolic, I've just pushed some ECS changes. But otherwise LGTM for a 2.x release!

**Comment by jeremynikolic** [comment]
Nice thanks @owenvoke 
I'll look into merging this one and adjusting the other with proper version 👌

**Comment by jeremynikolic** [comment]
@owenvoke if you may approve then 😄

**Review by owenvoke — APPROVED** [review]

**Comment by owenvoke** [comment]
Whoops my bad.

**Comment by owenvoke** [comment]
We'll also need to update the docs where necessary.

**Pull request merged**